### PR TITLE
Issue #13400 Create sectionedservice-test.txt

### LINKED ISSUE #13400: Write tests for services in sectionedservice
Go through the code, focusing on services, and write tests that traverses all paths in the code.
A step by step text file containing what actions to perform and evaluate the tests.

You should do thorough research of the logs, testing out different actions and what values are sent, if logging in or not changes behaviour. The models and functional requirements may also be applicable.

Follow the template in the README under `backend-models/service-tests/README.md`

See the process of work done in issue #13343!!!

Ignore "Retrieve Information"

### PR DESCRIPTION
**Pull request by c21alest**
V4 PR from old V3: #13528

Fixed #13400

### CODE REVIEW

File: backend-models/service-tests/sectionedservice-test.txt [hunk 1]
```diff
@@ -0,0 +1,551 @@
...
+getGroupsTest() line 100:
+----------------------------------
+
+*********
+ Pre-req
+*********
+
+(checklogin) IS TRUE
+
+//// PRE-REQ Values /////
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:13]: I'm a bit confused are, the sent values meant to be part of the pre-req or is there no pre-req?

File: backend-models/service-tests/sectionedservice-test.txt [hunk 2]
```diff
@@ -0,0 +1,551 @@
...
+$opt = "GRP"
+$courseid = "1"
+$coursevers = "45656"
+}
+
+/////////////////////////
+
+
+*****************
+ Compare results
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:26]: No comparisons are to be made in this doc, only input and output are to be documented (and some times some pre-req data need to be created in the database)

File: backend-models/service-tests/sectionedservice-test.txt [hunk 3]
```diff
@@ -0,0 +1,551 @@
...
+*****************
+
+If $debug is empty then successfully gathered all groups
+
+*****************
+ Service output
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:32]: the output that is to be recorded are the values in this array 
image
image

File: backend-models/service-tests/sectionedservice-test.txt [hunk 4]
```diff
@@ -0,0 +1,551 @@
...
+
+(checklogin) IS TRUE
+AND $ha || $studentTeacher IS TRUE
+
+//// PRE-REQ Values /////
+
+send
+{
+$opt = "DEL"
+$sectid != null
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:58]: it is not enough to just say that the variable should not have the value NULL.
an actual value needs to be specified.

File: backend-models/service-tests/sectionedservice-test.txt [hunk 5]
```diff
@@ -0,0 +1,551 @@
...
+}
+
+/////////////////////////
+
+You also need to gather all listentries to compare if deleted.
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:63]: no comparison in this doc

File: backend-models/service-tests/sectionedservice-test.txt [hunk 6]
```diff
@@ -0,0 +1,551 @@
...
+*****************
+ Compare results
+*****************
+
+If beforeDeleted(count) > afterDeleted(count): passed
+If beforeDeleted(count) < afterDeleted(count): failed
+If $debug not empty: failed
+
+*****************
+ Service output
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:81]: specify output. should be the same as first test.

File: backend-models/service-tests/sectionedservice-test.txt [hunk 7]
```diff
@@ -0,0 +1,551 @@
...
+*****************
+
+"codeexamples"
+
+----------------------------------
+deletedTest() line 197:
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:87]: this test has the same problems as deleteTest() line 181:

File: backend-models/service-tests/sectionedservice-test.txt [hunk 8]
```diff
@@ -0,0 +1,551 @@
...
+*********
+ Pre-req
+*********
+
+(checklogin) IS TRUE
+AND $ha || $studentTeacher IS TRUE
+
+//// PRE-REQ Values /////
+
+send
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:143]: all attributes should have a value, just stating that they are not null is not enough.
the person writing the actual test should not have to figure out what values can be used. Those values are to be documented here.

File: backend-models/service-tests/sectionedservice-test.txt [hunk 9]
```diff
@@ -0,0 +1,551 @@
...
+$kind != null
+$comments != null
+$visibility != null
+$highscoremode != null
+$pos != null
+}
+
+/////////////////////////
+
+You also need to gather all codeexamples to compare if any new.
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:164]: again no comparisons in this doc

File: backend-models/service-tests/sectionedservice-test.txt [hunk 10]
```diff
@@ -0,0 +1,551 @@
...
+*****************
+ Compare results
+*****************
+
+If any new table to codeexample and listentries: passed
+If non new table to codeexample and listentries: failed
+If $debug not empty: failed
+
+*****************
+ Service output
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:185]: again same as first test.

File: backend-models/service-tests/sectionedservice-test.txt [hunk 11]
```diff
@@ -0,0 +1,551 @@
...
+*****************
+
+"codeexamples"
+
+----------------------------------
+newRecorderTest() line 275:
+----------------------------------
+
+*********
+ Pre-req
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:195]: separate pre-req and send

File: backend-models/service-tests/sectionedservice-test.txt [hunk 12]
```diff
@@ -0,0 +1,551 @@
...
+*********
+
+(checklogin) IS TRUE
+AND $ha || $studentTeacher IS TRUE
+
+//// PRE-REQ Values /////
+
+send
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:203]: values.. 
and this includes the array

File: backend-models/service-tests/sectionedservice-test.txt [hunk 13]
```diff
@@ -0,0 +1,551 @@
...
+{
+$opt = "RECORDER"
+$order != 0
+$armin[] != null
+}
+
+/////////////////////////
+
+You also need to gather all listentries to compare if any updated.
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:212]: no comparisons in this doc

File: backend-models/service-tests/sectionedservice-test.txt [hunk 14]
```diff
@@ -0,0 +1,551 @@
...
+
+If pos and moment has been updated: passed
+If non has been updated: failed
+If $debug not empty: failed
+
+*****************
+ Service output
+*****************
+
+Not able to identify output for listentries
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:233]: same as first test

File: backend-models/service-tests/sectionedservice-test.txt [hunk 15]
```diff
@@ -0,0 +1,551 @@
...
+
+----------------------------------
+updateTest() line 291:
+----------------------------------
+
+*********
+ Pre-req
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:240]: separate pre-req and send

File: backend-models/service-tests/sectionedservice-test.txt [hunk 16]
```diff
@@ -0,0 +1,551 @@
...
+*********
+
+(checklogin) IS TRUE
+AND $ha || $studentTeacher IS TRUE
+
+//// PRE-REQ Values /////
+
+send
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:248]: values~~~~~~

File: backend-models/service-tests/sectionedservice-test.txt [hunk 17]
```diff
@@ -0,0 +1,551 @@
...
+$feedbackquestion != null
+$moment != null
+$kind != null
+$link != null
+$visibility != 0
+}
+
+/////////////////////////
+
+You also need to gather all listentries to compare if any updated.
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:270]: ño comparisons in this doc

File: backend-models/service-tests/sectionedservice-test.txt [hunk 18]
```diff
@@ -0,0 +1,551 @@
...
+*****************
+ Compare results
+*****************
+
+If any new table to codeexample and list, and updated listentries: passed
+If non new table: failed
+If $debug not empty: failed
+
+*****************
+ Service output
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:294]: same as test 1

File: backend-models/service-tests/sectionedservice-test.txt [hunk 19]
```diff
@@ -0,0 +1,551 @@
...
+If beforeDeleted(count) > afterDeleted(count): passed
+If beforeDeleted(count) < afterDeleted(count): failed
+If $debug not empty: failed
+
+*****************
+ Service output
+*****************
+
+"codeexamples"
+
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:85]: When the test has been executed, the database is to be reverted to the state it was before.
i would recommend creating an insert that you later deleted trough this test. (or if the test fail by a query)

Comment (a21jeaha) [reply backend-models/service-tests/sectionedservice-test.txt:85]: this is true for ALL tests, therefore i will not write this in every test. you just need to make sure that this is true in every test.

File: backend-models/service-tests/sectionedservice-test.txt [hunk 20]
```diff
@@ -0,0 +1,551 @@
...
+----------------------------------
+updateDeadlineTest() line 371:
+----------------------------------
+
+(checklogin) IS TRUE
+AND $ha || $studentTeacher IS TRUE
+
+//// PRE-REQ Values /////
+
+send
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:308]: value VALUE value VALUE

File: backend-models/service-tests/sectionedservice-test.txt [hunk 21]
```diff
@@ -0,0 +1,551 @@
...
+
+If deadline, relativedeadline has been updated: passed
+If deadline, relativedeadline not updated (same as beforeUpdated): failed
+If $debug not empty: failed
+
+*****************
+ Service output
+*****************
+
+"codeexample"
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:339]: output, same as first test.

File: backend-models/service-tests/sectionedservice-test.txt [hunk 22]
```diff
@@ -0,0 +1,551 @@
...
+----------------------------------
+updateTabsTest() line 381:
+----------------------------------
+
+(checklogin) IS TRUE
+AND $ha || $studentTeacher IS TRUE
+
+//// PRE-REQ Values /////
+
+send
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:350]: separate pre-req and send

File: backend-models/service-tests/sectionedservice-test.txt [hunk 23]
```diff
@@ -0,0 +1,551 @@
...
+{
+$opt = "UPDATETABS"
+$sectid != 0
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:353]: vaaaaaaalues....

File: backend-models/service-tests/sectionedservice-test.txt [hunk 24]
```diff
@@ -0,0 +1,551 @@
...
+
+If gradesystem has been updated: passed
+If gradesystem not updated (same as beforeUpdated): failed
+If $debug not empty: failed
+
+*****************
+ Service output
+*****************
+
+Not able to identify output for listentries
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:380]: same as first test

File: backend-models/service-tests/sectionedservice-test.txt [hunk 25]
```diff
@@ -0,0 +1,551 @@
...
+
+----------------------------------
+updateVersTest() line 390:
```
Comment (a21jeaha) [inline backend-models/service-tests/sectionedservice-test.txt:383]: well.. it seems like the same problems repeat them selves, the same apply for the rest of the doc

**Review by a21jeaha — CHANGES_REQUESTED** [review]
this need rework.
there are practically no values assigned in the send section of the tests

**Comment by d21erisu** [comment]
i assigned b21ricni to the issue instead of c21alest since he is issuedude now.

File: backend-models/service-tests/sectionedservice-test.txt (hunk 1 above)
Comment (b21ricni) [reply backend-models/service-tests/sectionedservice-test.txt:13]: I presume the values are for the choose course that will be displayed

File: backend-models/service-tests/sectionedservice-test.txt (hunk 2 above)
Comment (b21ricni) [reply backend-models/service-tests/sectionedservice-test.txt:26]: Ah so Compared results should be removed and send values will be put here?

Comment (a21jeaha) [reply backend-models/service-tests/sectionedservice-test.txt:26]: the comparisons should be removed from the document, as that is not the information that these test are created to retrieve. only the output is of interest. 

what i meant with input was the parameters that are sent to the service, which should be documented under the section "send"

**Comment by c21alest** [comment]
This issue is now assigned to @b21ricni so closing this one. Can still view what needs to be changed in new file.

**Pull request closed without merging**